### PR TITLE
OIDC DEV UI - custom OIDC providers should appear on OIDC extension card and among provider tabs

### PR DESCRIPTION
**Pull request by michalvavrik**
Here https://quarkus.io/guides/security-openid-connect-dev-services#dev-services-and-ui-support-for-other-openid-connect-providers you can see that in the old DEV UI, it was possible for other extensions to add provider to OIDC card. Now it is not. This PR:

- makes it possible for any extension card to produce page for any other extension (namespace)
- rewrites relevant OIDC DEV UI docs section

@phillip-kruger problem is that when I currently just set namespace, than my custom extension page appears on extension list on custom extension card, however when I go to OIDC Keycloak provider page, I can see on the right-most tabs that there is both custom provider (from custom extension) and Keycloak provider. That is strange. However if you generally don't want to allow extensions to place its pages to other extension cards, we can create in OIDC extension build item that will enable it only for OIDC. WDYT?

/cc @sberyozkin but please don't review docs text till there is an agreement on how to do this (so that you don't waste your time!)

custom-provider-card
custom-provider-tabs

### CODE REVIEW

**Comment by phillip-kruger** [comment]
@michalvavrik I do not follow the issue ? How do you mean a custom card ?

**Comment by michalvavrik** [comment]
@phillip-kruger I used term card because when you go to `http://localhost:8080/q/dev-ui/extensions` and checkout HTML element, it has class card. I'm happy to go into detail, but I thought description was very clear. I can try to rephrase it, hope it is useful:

- see https://quarkus.io/guides/security-openid-connect-dev-services#dev-services-and-ui-support-for-other-openid-connect-providers extension like Google Security extension (this is fictional extension, I think this feature appears mainly for Rendarde, but we can't be certain as this is documented for a longer period now) can provide their own OIDC providers in addition to _built in_ like Keycloak; On the pictures above, you can see OKTA and in changed doc, there is documented way how did I do it (I tested it naturally)
- before this PR, when I set namespace to `io.quarkus.quarkus-oidc`, I can see on the right side there are tabs Keycloak provider, Keycloak Admin, OKTA, but when I go to `http://localhost:8080/q/dev-ui/extensions`, I can't see OKTA on OIDC card, but on my custom extension card instead; I think this behavior is not predictable, maybe I missed something in the docs?
- in order to keep current documented behavior and for custom extensions be able to provide their own OIDC providers to OIDC namespace, I made this PR. As I explained in the PR desc, there are other options, please share your POV

Please let me know your questions and I'll do best to answer them.

**Comment by sberyozkin** [comment]
Hi @michalvavrik, let me clarify a few things.

Right now, in dev or prod mode, OIDC card will show `Okta` only, if `quarkus.oidc.auth-server-url=okta-address`.

This section you have linked to is about replacing the whole `Dev Services for Keycloak` with an alternative provider support in a zero config seup in devmode, so `quarkus.oidc.auth-server-url` is not set, we go to DevUI, and we only see `Okta` in the OIDC card, user clicks on `Okta` and sees something totally different to what current OIDC DevUI offers, something very Okta specific.

So I'm not quite sure yet what is the target of this PR ?

**Comment by michalvavrik** [comment]
Hey @sberyozkin @phillip-kruger your questions feels like there is something very wrong. It means I am missing something, but I don't know what (honestly!). I'll think about what I did wrong and come back to you.

Preliminary answer:

> Right now, in dev or prod mode, OIDC card will show Okta only, if quarkus.oidc.auth-server-url=okta-address.

It does not in new DEV UI (which is default). Correct? In old DEV UI, user provided https://github.com/quarkusio/quarkus/blob/main/extensions/oidc/deployment/src/main/resources/dev-templates/embedded.html#L1

```
            provider.produce(new DevConsoleRuntimeTemplateInfoBuildItem("io.quarkus", "quarkus-oidc", "oidcProviderName",
                    recorder.getProviderName()));
            provider.produce(new DevConsoleRuntimeTemplateInfoBuildItem("io.quarkus", "quarkus-oidc", "oidcProviderUrlBase",
                    recorder.getProviderUrlBase()));
```

we showed link to his provider on OIDC extension card. This is exactly what this appear does as well, except there is no `DevConsoleRuntimeTemplateInfoBuildItem` in new DEV UI, so when custom extension creates page, it sets correct namespace.

> This section you have linked to is about replacing the whole Dev Services for Keycloak with an alternative provider support in a zero config seup in devmode

We don't replace provider, you can have multiple OIDC providers. And conditions when custom extensions provides `oidcProviderName` and `oidcProviderUrlBase` were always at hands of custom extension, never at yours. You can see that `embedded.html` will accept it in case, when custom extension produces it. 

As for presence of `DevServicesConfigBuildItem`, I couldn't make sense of it, so I removed that **suggested** (but not enforced) condition. If you provide me a hint why it should be there, I'll also document it for users and add it back.

> , so quarkus.oidc.auth-server-url is not set, we go to DevUI, and we only see Okta in the OIDC card, user clicks on Okta and sees something totally different to what current OIDC DevUI offers, something very Okta specific.

I'm yet to test this, but it shouldn't be true. Okta link appears when custom extension provides these build items. I can't see any clear link to `quarkus.oidc.auth-server-url`. Can you help me understand what you mean, please?

**Comment by michalvavrik** [comment]
Maybe I understand what @sberyozkin tried to say -> so we should find a way to enforcer single OIDC provider?

**Comment by michalvavrik** [comment]
Now that I'm thinking about it, this is not in contradiction to changes here, maybe I mislead you in that that my pictures show Keycloak and OKTA, but that is because my dummy extension did not provide `quarkus.oidc.auth-server-url` and so, Keycloak was started.

I don't think it means my suggestions regarding Page namespace is wrong, is it?

**Comment by sberyozkin** [comment]
@michalvavrik You can configure that quickstart to have `quarkus.oidc.provider=google` and set some random client id/secret, in Dev UI, you will see a link to Google only in the OIDC card. Clicking on the link will fail since the client id/secret are random - but if you follow https://quarkus.io/guides/security-openid-connect-providers#google then you click on it and it just works the same way as for Keycloak - you sign in to Google and you will see Google ID token in JWT format, its binary access token.
That section is about replacing the current Dev UI experience with provider specific UI experience

**Comment by michalvavrik** [comment]
> @michalvavrik You can configure that quickstart to have `quarkus.oidc.provider=google` and set some random client id/secret, in Dev UI, you will see a link to Google only in the OIDC card. Clicking on the link will fail since the client id/secret are random - but if you follow https://quarkus.io/guides/security-openid-connect-providers#google then you click on it and it just works the same way as for Keycloak - you sign in to Google and you will see Google ID token in JWT format, its binary access token.

I'm sorry @sberyozkin , we completely and utterly don't understand each other.

> That section is about replacing the current Dev UI experience with provider specific UI experience

That is what this PR does, because that section is not valid for new DEV UI.

**Comment by sberyozkin** [comment]
>I'm sorry @sberyozkin , we completely and utterly don't understand each other.

Lol, it can happen sometimes :-)

>That is what this PR does, because that section is not valid for new DEV UI.

Right, so that was my question, what was the target of the PR. OK, I think I got it. 

What do the 2 images in PR description represent ?

**Comment by michalvavrik** [comment]
> What do the 2 images in PR description represent ?

I wanted to show that now, 

- custom extension can produce page for OKTA provider and show it on OIDC card at `http://localhost:8080/q/dev-ui/extensions`, problem is that I did not realize I should also produce `quarkus.oidc.auth-server-url` so that Keycloak is not started
- OKTA provider is shown on the tab among other providers; i context of ^^ it is not relevant for there will be just OKTA

There is unresolved question:

- way I allowed this feature is by setting namespace and you can see that I adjusted whole Vert.X HTTP because in my personal view behavior is strange (when I set namespace, OKTA is in OIDC tabs on the right, but on the card in `http://localhost:8080/q/dev-ui/extensions`
- we can avoid ^^ and simply create build item in OIDC deployment that will carry page and then we decide if we show it on OIDC card or not

**Comment by sberyozkin** [comment]
>custom extension can produce page for OKTA provider and show it on OIDC card at http://localhost:8080/q/dev-ui/extensions, problem is that I did not realize I should also produce quarkus.oidc.auth-server-url so that Keycloak is not started

You should not produce it - if one wants to provide some provider container support then there is a build item that Keycloak devservices processor checks at the very start indicating to it that it is not needed.

For Okta in particular, there is no container. For Okta `quarkus.oidc.auth-server-url` will have to be set in application.properties.

I wonder, if we can try a POC first and fix things along the way. Lets say someone would like a completely different OIDC Dev UI experience for `Keycloak` - they want to start a custom Keycloak container only (for POC it can be some specific KC image version), and they want to have OIDC card say: `Custom Keycloak` and the SPA page say `Custom Keycloak` (for POC existing provider.html can be copied) - so this POC has its own extension - empty runtime, deployment has this custom Keycloak support only. Now we add `quarkus-oidc`, we add this custom extension and `DevServices for Keycloak` in its current form is not visible anywhere.

That what I wanted to allow with that doc section - not exactly for custom Keycloak - but in general, to let users if they really want to, support zero config with some Provider X container, and also provide Provider X Dev UI , with zero config.

**Comment by michalvavrik** [comment]
> > custom extension can produce page for OKTA provider and show it on OIDC card at http://localhost:8080/q/dev-ui/extensions, problem is that I did not realize I should also produce quarkus.oidc.auth-server-url so that Keycloak is not started

Hmm, they cold probably try build time config source or config interceptor, but I get your point, extension is not going to do that, you put it to app properties.

> You should not produce it - if one wants to provide some provider container support then there is a build item that Keycloak devservices processor checks at the very start indicating to it that it is not needed.
> 
> For Okta in particular, there is no container. For Okta `quarkus.oidc.auth-server-url` will have to be set in application.properties.
> 

Make sense, I didn't provide it or realize there is only one possible provider, so it made images confusing.

> I wonder, if we can try a POC first and fix things along the way.

Hmm, I could write a test. I didn't go there to not over complicate it, but it can be done. We can do POC, but I'd need to understand expectations (what to discover, what to test), because I did try what I documented. Would you like me to create some git project with extension? I can do that, but such things gets obsolete over time when not tested / revised.

>  Lets say someone would like a completely different OIDC Dev UI experience for `Keycloak` - they want to start a custom Keycloak container only (for POC it can be some specific KC image version), and they want to have OIDC card say: `Custom Keycloak` and the SPA page say `Custom Keycloak` (for POC existing provider.html can be copied) - so this POC has its own extension - empty runtime, deployment has this custom Keycloak support only. Now we add `quarkus-oidc`, we add this custom extension and `DevServices for Keycloak` in its current form is not visible anywhere.

I see, so we should separate actual dev service started (container) from card/page visible in DEV UI. Hmm, that will require different solution.

> That what I wanted to allow with that doc section - not exactly for custom Keycloak - but in general, to let users if they really want to, support zero config with some Provider X container, and also provide Provider X Dev UI , with zero config.

Currently the doc section only allows you to customize DEV UI, it is not related to the container itself, correct? However OIDC won't start containers when auth URL is provided (and few other conditions).

I believe now it is clearer to me: **we should only produce OIDC page (and link on the OIDC extension card) when custom extension did not, but we should still start container dev svc (container) and for example setup KC unless auth URL is provided**. @sberyozkin if that is correct, I'll change this PR.

---------------------------------------------------
@phillip-kruger please allow me to say my opinion regarding namespace (see changed code in this PR): it seems like a bug and I'd like your take on it as it has impact on options we have. Thank you

**Comment by sberyozkin** [comment]
Hi @michalvavrik Sorry if I'm not exactly following, let me try again. 

So that doc section was just an idea spoken aloud - I was thinking, what if uses don't use Keycloak, but their provider has a container support, so I typed that doc section suggesting how `Dev Services for Keycloak` can be totally replaced - I did not have a clear idea if it would work - I just wanted to offer that idea to explore if someone would be interested.

Now, I haven't heard a single question about that section - which nearly 100% means no one has tried it

So, your PR may be good, but I'm not quite sure what exactly we are fixing since we don't have a single case supporting that idea. 

Right, we want `Okta`/etc shown - why alongside other providers - I don't now (even though you've clarified in this case it would be Okta only). What happens after that we have no clear idea.

This is why I thought it would be useful for both of us to have a simple POC based on Keycloak to make it simpler to discuss. 

This is case 1, Provider X offers their own container, OIDC devservice configuration, and OIDC Dev UI

And this is case 2: if `quarkus.oidc.auth-server-url` point for ex to Auth0, users do not want to use current OIDC Dev UI - but they want to offer Auth0 centric Dev UI with more specifics about tokens, etc.

IMHO, it makes sense, as without any POC, it is difficult to see what this PR really changes apart from fixing what looks like an obvious problem with `Okta`/etc link not being shown in new Dev UI in this case - but I can't check this PR against anything.

Hope it clarifies my concern, like I said, it may be a very precise, good PR as it is, but it is hard to verify it.

Thanks

**Comment by sberyozkin** [comment]
I think if we have POC and it can show Okta and other provider users how they can actually accomplish a task like that, in general it would very nice if users can do that

**Comment by michalvavrik** [comment]
Makes sense. I'll create POC and make this draft for the moment being. Also removing @phillip-kruger (please let me know your answer when you happen to have a time, thank you) as this seems like OIDC PR.

**Comment by sberyozkin** [comment]
Sounds good @michalvavrik, you can just try replace current OIDC Dev UI with Google (simple to setup), it does not have to be production ready replacement :-), copying and pasting existing OIDC DevUI template with some basic changes will do, or whatever you prefer :-)

**Comment by michalvavrik** [comment]
Created POC https://github.com/michalvavrik/custom-oidc-provider-devui-page-poc/ but so far there is no use to it. It is not clear to me how we can achieve what we need without changed in vertx-http dev ui. I pinged @phillip-kruger on Zulip to gather feedback.
